enhancements/cluster-logging/cluster-logging-log-forwarding.md

```diff
@@ -104,198 +156,160 @@ This is often required for industries such as the US public sector, healthcare o
+     secret:
+        name: my_secrets # Must contain keys tls.key, tls.cert and ca.cert 
+
+  pipelines:
+   - outputs: [ secure-tool ]
 ```
 
-Following is a diagram which identifies the workflow of a message through fluentd as captured by the above configuration.
+#### As a cluster administrator, I want to clearly separate where the logging stack forwards infrastructure and/or audit related logs.
+
+```
+apiVersion: "logging.openshift.io/v1alpha1"
+kind: "Forwarder"
+metadata:
+  namespace: cluster-logging # Global forwarder
+  pipelines:
+   - selector: { source: [Infrastructure] }
+     outputs: [ elasticsearch+https://infra.cluster.local ]
+   - selector: { source: [Application] }
+     outputs: [ elasticsearch+https://app.cluster.local ]
+   - selector: { source: [Audit] }
+     outputs: [ syslog+tls://mysyslogservice.example.net ]
+```
 
-![https://docs.google.com/drawings/d/18FD4c1pBz4e8eJIH37yUwM40ew69g8h1cqPEfOLsuyY/edit?usp=sharing](images/log-forwarding-message-flow.png)
+#### As a Red Hat SRE who operates OSD clusters, I want a mechanism to protect my configuration (e.g. audit log forwarding, infra logs) from non SRE administrators  of OSD but at the same time give them the opportunity to configure their own log forwarding for applications.
+
+Create global forwarder with RBAC control restricted to SRE admins. Users are free to create their own forwarders (global or namespace-restricted) independent of the SRE forwarder.
+
+```
+apiVersion: "logging.openshift.io/v1alpha1"
+kind: "Forwarder"
+metadata:
+  namespace: cluster-logging # Global forwarder
+spec:
+   - selector: { source: [Audit, Infrastructure] }
+     outputs: [ elasticsearch+https://sre.destination.example.com ]
+```
+
+#### I want to configure log forwarding to include/exclude logs on k8s labels.
+
+This applies equally to global or namespace-restricted forwarders.
+
+```
+apiVersion: "logging.openshift.io/v1alpha1"
+kind: "Forwarder"
+spec:
+  outputs: ... define outputs called 'everything' and 'critical'
+  pipelines:
+   - outputs: [ everything ] # All logs.
+   - selector: { labels: { importance: critical } }
+     outputs: [ critical ] # Only logs from containers labeled importance:critical.
+```
+
+#### As a cluster administrator, I want to configure a safety net if the remote service is not available so that I do not lose any of the logs.
+
+Send logs to both primary and backup:
+```
+...
+spec:
+  outputs: ...
+  pipelines:
+   - outputs: [ primary, backup ]
+```
 
 ### Cluster Logging Operator
-* The operator generates output configuration for the collector
+* The operator generates output configuration for the collector from all forwarder CRs.
 * The operator specifies in the collector daemonset the secrets needed to communicate to each endpoint:
 
-   `/var/run/ocp-collector/secrets/<secretName>/{tls.crt,tls.key,ca-bundle.crt}`
+   `/var/run/ocp-collector/secrets/<namespace>/<secretName>/{tls.crt,tls.key,ca-bundle.crt,username,password}`
```

Comment:
FieldSelector - ARRgh! That's something else altogether. It does look like this API has abandoned map selectors though, and only provides expresssions, so that's a start. However k8s dumb-ass selector syntax can't express logical OR, so if you want to be complete you need a list of such expressions. How can people with computer science degrees not know what a boolean expression is supposed to look?

Name and embedded union - yes that's (almost) where we ended up. A pipeline has a list of output names, each output has generic config and a union of supported type-specific config.

Global forwarders: are no longer in a namespace, they are cluster-scoped - seems more natural than a "magic" namespace. The CRD for namespace-scoped forwarders needs to be restricted and access controlled differently from cluster-scoped so it's not bad that it's a separte CRD - though the API will be nearly identical.  For GA we can store secrets the same way the TP does, we can work out the details for multiple forwarders later since' it's private to our code.